### PR TITLE
WDAPI-459 add new device posture rules

### PR DESCRIPTION
**Pull request by aplunk**
This adds support for domain_joined, firewall, os_version, and disk_encryption device posture rules.
DependsOn: https://github.com/cloudflare/cloudflare-go/pull/677

### CODE REVIEW

**Comment by aplunk** [comment]
What's the best way to run the acceptance tests?

File: website/docs/r/device_posture_rule.html.markdown [hunk 1]
```diff
@@ -69,6 +69,23 @@ The input structure depends on the device posture rule type.
...
+* `domain` = (Required) The domain that the client must join.
+
+**firewall**
+
+* `enabled` = (Required) True if the firewall must be enabled.
+
+**disk_encryption**
+
+* `require_all` = (Required) True if all drives must be encrypted.
+
```
Comment (jacobbednarz) [inline website/docs/r/device_posture_rule.html.markdown:88]: these descriptions don't match the schema states

```suggestion
* `version` = (Optional) The operating system semantic version.
* `operator` = (Optional) The version comparison operator in (>,>=,<,<=,==)

**domain_joined**

* `domain` = (Optional) The domain that the client must join.

**firewall**

* `enabled` = (Optional) True if the firewall must be enabled.

**disk_encryption**

* `require_all` = (Optional) True if all drives must be encrypted.
```

File: cloudflare/resource_cloudflare_device_posture_rule.go [hunk 2]
```diff
@@ -59,32 +59,64 @@ func resourceCloudflareDevicePostureRule() *schema.Resource {
...
-				Required: true,
+				Optional: true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"id": {
-							Type:     schema.TypeString,
-							Optional: true,
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "serial number rule: the id of the serial number list",
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_device_posture_rule.go:68]: Ideally, these descriptions should match the website documentation you've updated. In the future, the website documentation will be generated from this field. You can just include a description of the field itself (no need for things like `file rule:`, etc)

File: cloudflare/resource_cloudflare_device_posture_rule.go [hunk 3]
```diff
@@ -59,32 +59,64 @@ func resourceCloudflareDevicePostureRule() *schema.Resource {
...
+							Type:        schema.TypeBool,
+							Optional:    true,
+							Description: "file rule: true if the file must exist",
 						},
 						"thumbprint": {
-							Type:     schema.TypeString,
-							Optional: true,
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "application|file rule: the tumbprint of the binary",
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_device_posture_rule.go:83]:
```suggestion
							Description: "application|file rule: the thumbprint of the binary",
```

File: cloudflare/resource_cloudflare_device_posture_rule_test.go [hunk 4]
```diff
@@ -32,11 +32,58 @@ func TestAccCloudflareDevicePostureRuleBasic(t *testing.T) {
 					resource.TestCheckResourceAttr(name, "input.0.id", "asdf-123"),
 				),
 			},
+			{
+				Config: testAccCloudflareDevicePostureRuleConfigOsVersion(rnd, accountID),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_device_posture_rule_test.go:36]: can you please split these into individual tests? multiple assertions per test are very difficult to debug if something isn't correct in one of them but not the others.

Comment (jacobbednarz) [reply cloudflare/resource_cloudflare_device_posture_rule_test.go:36]: ["steps"](https://www.terraform.io/docs/extend/testing/acceptance-tests/teststep.html#steps) in Terraform acceptance tests are kind of like different states to test against. example is you create a resource and then update it N times to confirm the state.

File: go.mod [hunk 5]
```diff
@@ -3,7 +3,7 @@ module github.com/cloudflare/terraform-provider-cloudflare
 go 1.15
 
 require (
-	github.com/cloudflare/cloudflare-go v0.19.0
+	github.com/cloudflare/cloudflare-go v0.19.1-0.20210728051236-7210b1583143
```
Comment (jacobbednarz) [inline go.mod:6]: can you please remove this from your PR? we bump this separately to the PR.

File: website/docs/r/device_posture_rule.html.markdown (hunk 1 above)
Comment (aplunk) [reply website/docs/r/device_posture_rule.html.markdown:88]: The documentation is correct, the validation is too lax. I'm not sure how to perform validation for the *input* field when it's dependent on the *type* field.

File: cloudflare/resource_cloudflare_device_posture_rule.go (hunk 2 above)
Comment (aplunk) [reply cloudflare/resource_cloudflare_device_posture_rule.go:68]: I'll update these.

File: cloudflare/resource_cloudflare_device_posture_rule.go (hunk 3 above)
Comment (aplunk) [reply cloudflare/resource_cloudflare_device_posture_rule.go:83]: Good catch.

File: cloudflare/resource_cloudflare_device_posture_rule_test.go (hunk 4 above)
Comment (aplunk) [reply cloudflare/resource_cloudflare_device_posture_rule_test.go:36]: I will make this change.

File: go.mod (hunk 5 above)
Comment (aplunk) [reply go.mod:6]: Will do 👍

File: website/docs/r/device_posture_rule.html.markdown (hunk 1 above)
Comment (jacobbednarz) [reply website/docs/r/device_posture_rule.html.markdown:88]: you can't really do it; usually the validation goes in the CRUD operations themselves and these fields are marked as optional from the schema point of view.

**Review by jacobbednarz — APPROVED** [review]
we're looking good on the acceptance testing front

```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./...) -v -run "^TestAccCloudflareDevicePostureRule" -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestAccCloudflareDevicePostureRuleSerialNumber
--- PASS: TestAccCloudflareDevicePostureRuleSerialNumber (6.02s)
=== RUN   TestAccCloudflareDevicePostureRuleOsVersion
--- PASS: TestAccCloudflareDevicePostureRuleOsVersion (5.70s)
=== RUN   TestAccCloudflareDevicePostureRuleDomainJoined
--- PASS: TestAccCloudflareDevicePostureRuleDomainJoined (18.05s)
=== RUN   TestAccCloudflareDevicePostureRuleFirewall
--- PASS: TestAccCloudflareDevicePostureRuleFirewall (6.02s)
=== RUN   TestAccCloudflareDevicePostureRuleDiskEncryption
--- PASS: TestAccCloudflareDevicePostureRuleDiskEncryption (5.56s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	42.089s
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
```

just waiting on the cloudflare-go release next week to land this

File: website/docs/r/device_posture_rule.html.markdown (hunk 1 above)
Comment (aplunk) [reply website/docs/r/device_posture_rule.html.markdown:88]: How do you feel about updating the crud operations validation to include something like this?

```
func setDevicePostureRuleInput(rule *cloudflare.DevicePostureRule, d *schema.ResourceData) error {
	if _, ok := d.GetOk("input"); ok {
		input := cloudflare.DevicePostureRuleInput{}

		switch rule.Type {
		case "file":
			p, ok := d.GetOk("input.0.path")
			if !ok {
				return fmt.Errorf("file rule requires input.path")
			}
			input.Path = p.(string)

			exists, ok := d.GetOk("input.0.exists")
			if !ok {
				// File rule exists to true.
				exists = true
			}
			input.Exists = exists.(bool)

			if tp, ok := d.GetOk("input.0.thumbprint"); ok {
				input.Thumbprint = tp.(string)
			}

			if s, ok := d.GetOk("input.0.sha256"); ok {
				input.Sha256 = s.(string)
			}
		case "application":
			p, ok := d.GetOk("input.0.path")
			if !ok {
				return fmt.Errorf("application rule requires input.path")
			}
			input.Path = p.(string)

			if tp, ok := d.GetOk("input.0.thumbprint"); ok {
				input.Thumbprint = tp.(string)
			}

			if s, ok := d.GetOk("input.0.sha256"); ok {
				input.Sha256 = s.(string)
			}

			running, ok := d.GetOk("input.0.running")
			if !ok {
				// Application rule defaults running to true.
				running = true
			}
			input.Running = running.(bool)
		case "serial_number":
			inputID, ok := d.GetOk("input.0.id")
			if !ok {
				return fmt.Errorf("serial_number rule requires input.id")
			}
			input.ID = inputID.(string)
		case "disk_encryption":
			require_all, ok := d.GetOk("input.0.require_all")
			if !ok {
				// Disk encryption rule defaults enabled to true.
				require_all = false
			}
			input.RequireAll = require_all.(bool)
		case "firewall":
			enabled, ok := d.GetOk("input.0.enabled")
			if !ok {
				// Firewall rule defaults enabled to true.
				enabled = true
			}
			input.Enabled = enabled.(bool)
		case "os_version":
			if version, ok := d.GetOk("input.0.version"); ok {
				input.Version = version.(string)
			}
			if operator, ok := d.GetOk("input.0.operator"); ok {
				input.Operator = operator.(string)
			}
		case "domain_joined":
			domain, ok := d.GetOk("input.0.domain")
			if !ok {
				return fmt.Errorf("domain rule requires input.domain")
			}
			input.Domain = domain.(string)
		}

		rule.Input = input
	}

	return nil
}
```

Comment (jacobbednarz) [reply website/docs/r/device_posture_rule.html.markdown:88]: we generally use documentation and examples to describe the correct configuration and leave the more complex validation to the API itself. this avoids duplication in multiple systems such as here and `cf-terraforming` and it's very difficult to keep it all in sync correctly.

**Review by jacobbednarz — APPROVED** [review]
acceptance tests are all good!

```
TF_ACC=1 go test $(go list ./...) -v -run "^TestAccCloudflareDevicePosture" -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestAccCloudflareDevicePostureRuleSerialNumber
--- PASS: TestAccCloudflareDevicePostureRuleSerialNumber (7.17s)
=== RUN   TestAccCloudflareDevicePostureRuleOsVersion
--- PASS: TestAccCloudflareDevicePostureRuleOsVersion (6.33s)
=== RUN   TestAccCloudflareDevicePostureRuleDomainJoined
--- PASS: TestAccCloudflareDevicePostureRuleDomainJoined (5.88s)
=== RUN   TestAccCloudflareDevicePostureRuleFirewall
--- PASS: TestAccCloudflareDevicePostureRuleFirewall (6.81s)
=== RUN   TestAccCloudflareDevicePostureRuleDiskEncryption
--- PASS: TestAccCloudflareDevicePostureRuleDiskEncryption (6.89s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	33.743s
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
```

let's ship! thanks for this @aplunk, you rock 🤘

**Comment by jacobbednarz** [comment]
rebased this one from master and re-ran the tests. still looking good! let's merge.

**Pull request merged**